spring-integration-core/src/main/java/org/springframework/integration/aggregator/ResequencingMessageHandler.java

```diff
@@ -81,13 +83,21 @@ protected void afterRelease(MessageGroup messageGroup, Collection<Message<?>> co
 			remove(messageGroup);
 		}
 		else {
-			if (completedMessages != null){
-				int lastReleasedSequenceNumber = this.findLastReleasedSequenceNumber(messageGroup.getGroupId(), completedMessages);
-				messageStore.setLastReleasedSequenceNumberForGroup(messageGroup.getGroupId(), lastReleasedSequenceNumber);
-				this.messageStore.removeMessagesFromGroup(messageGroup.getGroupId(), completedMessages);
+			Object groupId = messageGroup.getGroupId();
+			if (completedMessages != null) {
+				int lastReleasedSequenceNumber = findLastReleasedSequenceNumber(groupId, completedMessages);
+				this.messageStore.setLastReleasedSequenceNumberForGroup(groupId, lastReleasedSequenceNumber);
+				if (this.messageStore instanceof SimpleMessageStore
+						&& completedMessages.size() == messageGroup.size()) {
+					((SimpleMessageGroup) messageGroup).clear();
+					((SimpleMessageGroup) messageGroup).setLastModified(System.currentTimeMillis());
```

Comment:
Don't see reason. we need the "hard" clean only for the `SMS` with `UpperBound`.
All other cases we can consider as new feature and really can address with the new point release.